### DISCUSSION TITLE
Question on how to make and use library.

### DISCUSSION THREAD

**Discussion by cloudhan** [gpuweb/gpuweb#2145] [Q&A] [answered]
I have been reading WebGPU Shading Language spec these days. And spec is pretty well writen. Well done.

However, the language is is not that strong enough IMHO. Seems there is no built-in langauge feature to let developer maintain a seperate wgsl library for common tasks.

Take [cub](https://github.com/NVIDIA/cub) as an example, I'd imagine those cooperative parallel primitives will be widely used (in terms of compute). Is there any way to form a cub like library and bring them into user src?

That is, how to do `#include` like in C++ or `mod` like in rust at language level.

**Comment by TUSF** [comment]
This is because WGSL is designed to trivially map to SPIR-V, which as far as I'm aware does not have such library features either. Instead, you would have a singular Shader module with all your needed tasks and multiple entry points, and a Pipeline that uses one of said entry-points.

For maintaining libraries however, it might not be a bad idea to have something akin to compiler directives, such as an `import` directive, to a library into the current module's storage space. Probably with explicitly declared namespace prefixes.

**Reply by kainino0x** [reply]
SPIR-V does have a concept of "linking" to link shader modules together with C-style forward declarations. However it's TBD how we would achieve useful/usable code sharing in WGSL e.g. linking multiple GPUShaderModules together in a single pipeline.

**Comment by Kangz** [comment, marked as answer]
Another aspect is that to make a library like cub useful, Nvidia used heavy generic programming with C++ templates. At the moment there is no great way to make library that can be used in a generic fashion, however WGSL is still improving and your feedback is IMHO pretty useful in that it shows something that's lacking in WGSL today.
